### PR TITLE
[Datasets] Add local and S3 filesystem test coverage for file-based datasources.

### LINKED ISSUE #17084: [Dataset] Add S3-backed tests for Ray Dataset IO layer.
We should cover usage of the S3 Arrow filesystem in our Ray Dataset IO layer tests to ensure that we're not e.g. closing over unserializable objects in our read/write tasks.

### PR DESCRIPTION
**Pull request by clarkzinzow**
Ensures that manually specified local and S3 filesystems are covered in our file-based datasources tests.

## Related issue number

Closes #17084

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/data/datasource/file_based_datasource.py [hunk 1]
```diff
@@ -169,6 +173,23 @@ def _resolve_paths_and_filesystem(
     return expanded_paths, file_infos, filesystem
 
 
+S3_PROTOCOL = "s3://"
+S3A_PROTOCOL = "s3a://"
+S3N_PROTOCOL = "s3n://"
+PROTOCOLS = [S3_PROTOCOL, S3A_PROTOCOL, S3N_PROTOCOL]
+
+
+def _unwrap_protocol(path):
```
Comment (ericl) [inline python/ray/experimental/data/datasource/file_based_datasource.py:182]: Can you use URI parsing here instead?

File: python/ray/experimental/data/tests/test_dataset.py [hunk 2]
```diff
@@ -913,88 +917,140 @@ def test_to_torch_feature_columns(ray_start_regular_shared):
     assert np.array_equal(df.values, combined_iterations)
 
 
-def test_json_read(ray_start_regular_shared, tmp_path):
+@pytest.mark.parametrize("fs,data_path,endpoint_url", [
+    (None, lazy_fixture("local_path"), None),
+    (lazy_fixture("local_fs"), lazy_fixture("local_path"), None),
+    (lazy_fixture("s3_fs"), lazy_fixture("s3_path"), lazy_fixture("s3_server"))
+])
```
Comment (ericl) [inline python/ray/experimental/data/tests/test_dataset.py:924]: Nice.

**Review by ericl — APPROVED** [review]

File: python/ray/experimental/data/datasource/file_based_datasource.py [hunk 3]
```diff
@@ -144,6 +144,10 @@ def _resolve_paths_and_filesystem(
 
     resolved_paths = []
     for path in paths:
+        if filesystem is not None:
+            # If we provide a filesystem, _resolve_filesystem_and_path will not
+            # slice off the protocol from the provided URI/path when resolved.
+            path = _unwrap_protocol(path)
```
Comment (ericl) [inline python/ray/experimental/data/datasource/file_based_datasource.py:150]: Is this different from filesystem.normalize_path()?

Comment (clarkzinzow) [reply python/ray/experimental/data/datasource/file_based_datasource.py:151]: Unfortunately yes. that doesn't strip off the protocol:
```python
fs.normalize_path("s3://foo/bar") == "s3://foo/bar"
```
Meanwhile, `FileSystem.from_uri("s3://foo/bar)` will return `"foo/bar"`, along with a newly constructed `S3FileSystem`, which seemed wasteful.

**Review by wuisawesome — APPROVED** [review]

**Review by ericl — DISMISSED** [review]
We need to add parquet tests here (currently S3 parquet read support is totally broken since we don't import pyarrow.fs before deserialization of the filesystem impl).

**Comment by clarkzinzow** [comment]
Still hitting the infinite recursion here when pickling. ☹️

**Comment by ericl** [comment]
Can you ask Yi for help on ideas in resolving this problem?

**Comment by clarkzinzow** [comment]
CC @iycheng @suquark, also going to ping y'all in the Slack thread that has more info.

**Comment by clarkzinzow** [comment]
Datasets tests are looking good! This should be ready to merge.

**Comment by ericl** [comment]
🎉

**Comment by hckuo** [comment]
This PR adds back `opencv-python-headless` as a req package which was removed here. #16929

**Pull request merged**